src/openvic-simulation/misc/Define.cpp

```diff
@@ -1,109 +1,249 @@
 
 using namespace OpenVic;
 using namespace OpenVic::NodeTools;
 
-Define::Define(std::string_view new_identifier, std::string&& new_value, Type new_type)
-	: HasIdentifier { new_identifier }, value { std::move(new_value) }, type { new_type } {}
+std::string_view Define::type_to_string(Type type) {
+	using enum Type;
+
+	switch (type) {
+	case Date: return "date";
+	case Country: return "country";
+	case Economy: return "economy";
+	case Military: return "military";
+	case Diplomacy: return "diplomacy";
+	case Pops: return "pops";
+	case Ai: return "ai";
+	case Graphics: return "graphics";
+	case Unknown: [[fallthrough]];
+	default: return "unknown";
+	}
+}
+
+Define::Type Define::string_to_type(std::string_view str) {
+	using enum Type;
 
-fixed_point_t Define::get_value_as_fp() const {
-	return fixed_point_t::parse(value);
+	static const string_map_t<Define::Type> type_map {
+		{ "country",   Country },
+		{ "economy",   Economy },
+		{ "military",  Military },
+		{ "diplomacy", Diplomacy },
+		{ "pops",      Pops },
+		{ "ai",        Ai },
+		{ "graphics",  Graphics },
+	};
+
+	const string_map_t<Define::Type>::const_iterator type_it = type_map.find(str);
+
+	if (type_it != type_map.end()) {
+		return type_it->second;
+	} else {
+		return Unknown;
+	}
 }
 
-int64_t Define::get_value_as_int() const {
-	return std::strtoll(value.data(), nullptr, 10);
+Define::Define(std::string_view new_identifier, std::string_view new_value, Type new_type)
+	: HasIdentifier { new_identifier }, value { new_value }, type { new_type } {}
+
+Date Define::get_value_as_date(bool* successful) const {
+	return Date::from_string(value, successful);
 }
 
-uint64_t Define::get_value_as_uint() const {
-	return std::strtoull(value.data(), nullptr, 10);
+fixed_point_t Define::get_value_as_fp(bool* successful) const {
+	return fixed_point_t::parse(value, successful);
 }
 
-bool DefineManager::add_define(std::string_view name, std::string&& value, Define::Type type) {
-	if (name.empty()) {
-		Logger::error("Invalid define identifier - empty!");
-		return false;
-	}
-	return defines.add_item({ name, std::move(value), type }, duplicate_warning_callback);
+int64_t Define::get_value_as_int(bool* successful) const {
+	return StringUtils::string_to_int64(value, successful);
 }
 
-Date DefineManager::get_start_date() const {
-	return start_date ? *start_date : Date {};
+uint64_t Define::get_value_as_uint(bool* successful) const {
+	return StringUtils::string_to_uint64(value, successful);
 }
 
-Date DefineManager::get_end_date() const {
-	return end_date ? *end_date : Date {};
+std::ostream& OpenVic::operator<<(std::ostream& os, Define::Type type) {
+	return os << Define::type_to_string(type);
 }
 
-bool DefineManager::in_game_period(Date date) const {
-	if (start_date && end_date) {
-		return date.in_range(*start_date, *end_date);
+template<typename T>
+bool DefineManager::load_define(T& value, Define::Type type, std::string_view name) const {
```

Comment:
wouldn't optional<T> be more ergonomic here? Using an output parameter that is *maybe* set is error prone because it will be easy to forget to check the return value (unless you add `[[nodiscard]]`) and try to use a potentially uninitialized value.

Reply:
The return value is only used to contribute to `load_defines_file`'s return value, the define variables are default initialised in `DefineManager`'s constructor so no further fallback is required if `load_define` fails.